modules/table/src/main/java/org/apache/ignite/internal/table/TupleBuilderImpl.java

```diff
@@ -89,72 +91,164 @@ public TupleBuilder set(Column col, Object value) {
     }
 
     /** {@inheritDoc} */
-    @Override public <T> T valueOrDefault(String colName, T def) {
-        return (T)map.getOrDefault(colName, def);
+    @Override public String columnName(int columnIndex) {
+        return schemaDesc.column(columnIndex).name();
     }
 
     /** {@inheritDoc} */
-    @Override public <T> T value(String colName) {
-        return (T)map.get(colName);
+    @Override public Integer columnIndex(String columnName) {
+        var col = schemaDesc.column(columnName);
+
+        return col == null ? null : col.schemaIndex();
```

Comment:
'Null here breaks the contract.
Assume, user pass a value for a column that do not exists in the schema yet,
then it is impossible to get an index for this column.

Tuple interface says 'null' index will be returned if column doesn't exists in the Tuple,
but it is not true, obviously.

I think schema columns order and columns ID is an internal part, thus
either access methods by id shouldn't be exposed to user (to public API),
or column index should be a synthetic.

In general, Tuple may miss values for some columns (similar to Ignite 2).
What does column index means in that case?

Reply:
We use schema in TupleBuilderImpl just for optimizations.
As we already on a server, then we have a schema and we can validate columns instantly on set operation,
and on happy-path avoid unwanted Map (col name -> value) copying.

Reply:
We don't expose schemas to the user. Let's abstract away from schemas.

Every Tuple is a set of values. Values can be accessed by name and by index. It provides a way to access Tuple contents efficiently, without hash lookups, e.g.:

```
for (int i = 0; i < tuple.columnCount(); i++) {
    System.out.println(tuple.columnName(i) + ": " + tuple.value(i));
} 
```

There is nothing in the new API that requires a schema underneath. For example, client-side Live Schema Tuple can be backed by a `LinkedHashMap` without any prior schema knowledge.

Reply:
> In general, Tuple may miss values for some columns (similar to Ignite 2).
> What does column index means in that case?

When column is missing, `columnIndex` returns `null`.

Reply:
> Tuple interface says 'null' index will be returned if column doesn't exists in the Tuple,
> but it is not true, obviously.

Can you please clarify?
Do you mean that schema-based tuple will return non-null index even if column value is not set?

Reply:
Yes, make sense.

I mean, what if user set non-null value for a column that doesn't exists in schema (Live-schema case) and then can't find it by index in the Tuple.
As you answered earlier, this will be covered in the next PR.

Reply:
@AMashenkov I've updated `ClientTupleBuilder` to differentiate unset and null values, so it behaves the same as `TupleBuilderImpl` on the server. This behavior is covered by `ClientTupleBuilderTest` and `TupleBuilderImplTest`, which have the same tests for different implementations.